pkg/R/utils.R

```diff
@@ -288,3 +254,117 @@ convertEnvsToList <- function(keys, vals) {
+processClosure <- function(node, oldEnv, argNames, newEnv) {
+  nodeLen <- length(node)
+  if (nodeLen == 0) {
+    return
+  }
+  if (nodeLen > 1 && typeof(node) == "language") {
+    # Recursive case: current AST node is an internal node, check for its children. 
+    if (length(node[[1]]) > 1) {
+      for (i in 1:nodeLen) {
+        processClosure(node[[i]], oldEnv, argNames, newEnv)
+      }
+    } else {  # if node[[1]] is length of 1, might be an R primitive.
+      nodeChar <- as.character(node[[1]])
+      switch(nodeChar, 
+             "{" = {  # Start of a function body.
+               for (i in 2:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }
+             },
+             "<-" = {  # Assignment.
+               defVar <- node[[2]]
+               if (length(defVar) == 1 && typeof(defVar) == "symbol") {
+                 # Add the defined variable name into .defVars.
+                 assign(".defVars", 
+                        c(get(".defVars", envir = .sparkREnv), as.character(defVar)), 
+                        envir = .sparkREnv)
+               }
+               for (i in 3:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }
+             },
+             "function" = {  # Function definition.
+               newArgs <- names(node[[2]])
+               argNames <- c(argNames, newArgs)  # Add parameter names.
+               for (i in 3:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }
+             }, 
+             {
+               for (i in 1:nodeLen) {
+                 processClosure(node[[i]], oldEnv, argNames, newEnv)
+               }             
+            })
+    }
+  } else if (nodeLen == 1 && 
+               (typeof(node) == "symbol" || typeof(node) == "language")) {
+    # Base case: current AST node is a leaf node and a symbol or a function call.
+    nodeChar <- as.character(node)
+    if (!nodeChar %in% argNames && # Not a function parameter or function local variable.
+          !nodeChar %in% get(".defVars", envir = .sparkREnv)) {
+      func.env <- oldEnv
+      topEnv <- parent.env(.GlobalEnv)
+      # Search in function environment, and function's enclosing environments 
+      # up to global environment. There is no need to look into package environments
+      # above the global or namespace environment that is not SparkR below the global, 
+      # as they are assumed to be loaded on workers.
+      while (!identical(func.env, topEnv)) {
+        # Namespaces other than "SparkR" will not be searched.
+        if (!isNamespace(func.env) || getNamespaceName(func.env) == "SparkR") {
+          # Set parameter 'inherits' to FALSE since we do not need to search in
+          # attached package environments.
+          if (exists(nodeChar, envir = func.env, inherits = FALSE)) {
+            if (!isNamespace(func.env) || 
```

Comment:
Isn't this already ensured by the previous `if`. I am a bit confused as to why we need to check again.